com.unity.multiplayer.mlapi/Runtime/Core/NetworkTickSystem.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using UnityEngine;
+
+namespace MLAPI
+{
+    public class NetworkTickSystem : INetworkUpdateSystem, IDisposable
+    {
+        private const double k_DefaultTickDuration = 0.016;
+        private double m_TickDuration;
+        private int m_NetworkTick; //How many network ticks have passed?
+
+        private static NetworkTickSystem m_Instance = null;
+
+        // special value to indicate "No tick information"
+        public const ushort k_NoTick = ushort.MaxValue;
+        // Number of ticks over which the tick number wraps back to 0
+        public const ushort k_TickPeriod = k_NoTick - 1;
+
+        public static NetworkTickSystem Instance
+        {
+            get
+            {
+                if (m_Instance == null)
+                {
+                    m_Instance = new NetworkTickSystem();
+                }
+                return m_Instance;
+            }
+        }
+
+        public void Dispose()
+        {
+            NetworkUpdateLoop.UnregisterNetworkUpdate(this, NetworkUpdateStage.EarlyUpdate);
+        }
+
+        /// <summary>
+        /// GetTick
+        /// Gets the current network tick (non-fractional, wrapping around)
+        /// </summary>
+        /// <returns></returns>
+        public ushort GetTick()
+        {
+            return (ushort)(m_NetworkTick % k_TickPeriod);
+        }
+
+        /// <summary>
+        /// GetNetworkTime
+        /// Network time is calculated from m_NetworkTick and m_TickDuration (tick frequency)
+        /// </summary>
+        /// <returns>Network Time</returns>
+        public double GetNetworkTime()
+        {
+            return m_NetworkTick * m_TickDuration;
+        }
+
+        /// <summary>
+        /// UpdateNetworkTick
+        /// Called each network loop update during the PreUpdate stage
+        /// </summary>
+        private void UpdateNetworkTick()
+        {
+            m_NetworkTick = (int)(Time.unscaledTime / m_TickDuration);
```

Comment:
why `(int)(float / double)` instead of `(int)(float / float)`? `m_TickDuration` could be a `float` no?

Reply:
I had tried to keep the design for the original branch and thus kept the `double`s. But yeah, you raise a good point if our source data is `float`, no point going over-board with `double`. I've got in my plans to eventually review the float usage for precision. But `float` would give us already ~2ms resolution after 5hours game time. Good enough for now. Will change.

Reply:
We should keep all calculations as doubles (in my personal opinion).  The original experimental code I let run over night for about 10 hours and it didn't lose precision.

Reply:
Let's add a comment saying we plan to move to FixedTime, etc.